### PR TITLE
end-to-end tutorial for finetuning an LLM with the UI

### PR DESCRIPTION
**Pull request by burtenshaw**

### CODE REVIEW

File: ui-finetune/README.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+### Install the required libraries
+
+Before we can fine-tune the model, we need to install the required libraries. We will install `autotrain-advanced`, `datasets`, and `transformers`. We will also install the compatible versions of `torch`, `torchvision`, and `torchaudio`.
+
+```python
+# Uninstall existing torch, torchvision, and torchaudio
+pip uninstall torch torchvision torchaudio -y
+
+# Install compatible versions
+pip install torch==2.0.1+cu118 torchvision==0.15.2+cu118 torchaudio==2.0.2 --extra-index-url https://download.pytorch.org/whl/cu118
```
Comment (davidberenstein1957) [inline ui-finetune/README.md]: why do we need these and why do we need the cuda versions if we are going to use autotrain?

File: ui-finetune/README.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+# Install other required libraries
+pip install -U autotrain-advanced
+pip install -U datasets transformers
+```
+
+### Select the algorithm
```
Comment (davidberenstein1957) [inline ui-finetune/README.md:80]: doesn't autotrain just automatically use `DPO` at least you can't seem to configure this in the UI.

File: ui-finetune/README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+```bash
+huggingface_hub login
+python create_dataset.py \
+            --dataset_path argilla_dataset_name \
+            --dataset_repo_id your_hf_repo_id
+```
+
+## Fine-tune
+
+With a dataset on the hub, we can now fine-tune a language model using `autotrain-advanced`. We will use the `llm` command to fine-tune a language model on the dataset.
```
Comment (davidberenstein1957) [inline ui-finetune/README.md]: you don't actually show/use  the `llm` command, right?

File: ui-finetune/README.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+| weight-decay           | 0.01                | Weight decay for regularization              |
+| gradient-accumulation  | 4                   | Gradient accumulation steps                  |
+| mixed-precision        | bf16                | Mixed precision format (e.g., bf16, fp16)    |
+| max-prompt-length      | 256                 | Maximum length for prompts                   |
+| max-completion-length  | 256                 | Maximum length for completions               |
+| logging-steps          | 10                  | Steps interval for logging                   |
+| save-total-limit       | 2                   | Limit for total number of saved checkpoints  |
+| seed                   | 42                  | Random seed for reproducibility              |
+
+![alt text](../images/ui-finetune/autotrain_params.png)
```
Comment (davidberenstein1957) [inline ui-finetune/README.md:136]: it seems you are still using the original dataset and not the improved/exported one?

File: ui-finetune/README.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+- Once you've evaluated your model's checkpoints, you might wish to tweak epochs, weight-decay, and LoRa parameters.
+
+## Evaluate the model
+
+You can now evaluate your trained model. Here we will use some general benchmarks which can help to determine whether our model's performance has changed compared to its previous training. 
+
+For a real-world use case, you would want to to evaluate your model on the task that you plan to use it for. In this guide on [custom evaluation](domain-eval/README.md) we show how to do that. 
+
+```sh
+lighteval accelerate \
```
Comment (davidberenstein1957) [inline ui-finetune/README.md:147]: perhaps we can either add code that runs and do the correct installs or only redirect the user to the other blog?

File: ui-finetune/README.md (hunk 1 above)
Comment (davidberenstein1957) [reply ui-finetune/README.md]: Is this not included default by the UI/deployment?

File: ui-finetune/README.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+```python
+# Uninstall existing torch, torchvision, and torchaudio
+pip uninstall torch torchvision torchaudio -y
+
+# Install compatible versions
+pip install torch==2.0.1+cu118 torchvision==0.15.2+cu118 torchaudio==2.0.2 --extra-index-url https://download.pytorch.org/whl/cu118
+
+# Install other required libraries
+pip install -U autotrain-advanced
```
Comment (davidberenstein1957) [inline ui-finetune/README.md]: you don't seem to use the SDK anymore?

File: ui-finetune/README.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+### Filter the dataset
+
+After importing the dataset, we can filter it based on the relevance question we added. In our case, we will filter the dataset to only include completions that are relevant to the prompt. We will also filter the dataset to only include completions that are at least 10 tokens long.
+
+![filter dataset](../images/ui-finetune/filter_records.png)
+
+## Export the dataset to the Hub
+
+After filtering the dataset, we can export it to the hub. This will allow us to access the dataset in AutoTrain. Unfortunately, the export feature is not yet available in the UI, so we will have to use the python library to export the dataset.
```
Comment (davidberenstein1957) [inline ui-finetune/README.md:44]: given the complexity of the code (it is simple), i would just add it to the main file directly, especially since the blog tailors to people that would want to use easy UI and simple code to fine-tune, and not rely on external script calls.

File: ui-finetune/README.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+Developed by Hong and Lee in 2024, ORPO combines these processes by modifying the model’s objective function to include a loss term that rewards preferred responses while penalizing rejected ones. This approach has shown to outperform other alignment methods across different model sizes and tasks. ORPO’s efficiency and improved alignment make it a promising alternative in fine-tuning LLMs like Llama 3.
+
+### Select the base model
+
+The Hugging Face Hub contains thousands of language models that we could use as a base model for fine-tuning. We will use the SmolLM2 model as a base because it's only 1.7 billion parameters which means it will run on a wide range of hardware. It's also performed well on general benchmarks so we can expect reasonable performance from it on a number of use cases.
```
Comment (davidberenstein1957) [inline ui-finetune/README.md]: perhaps we can reference the llm leaderboard here?

File: ui-finetune/README.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+### Importing the dataset into Argilla
+
+To import the dataset into Argilla, we will use the 'Create Dataset' feature in the UI. For a more detailed guide on how to create a dataset in Argilla, check out this [blog post](https://huggingface.co/blog/argilla-ui-hub).
+
+The video below shows how to import the dataset into Argilla by pasting the dataset's repo id into the 'Create Dataset' form. In our case, the repo id is `mlabonne/orpo-dpo-mix-40k`.
+
+![video](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/argilla-ui-hub/import_hub_dataset.mp4)
+
+The dataset configurator lets you define a feedback task in the UI based on the dataset that you're importing. In our case, we will use the default task, and add a question for relevance. This will allow us to filter the dataset based on categories or topics.
```
Comment (sdiazlor) [inline ui-finetune/README.md]: For not argilla users, maybe talking about feedback tasks is not so obvious.

File: ui-finetune/README.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+![alt text](../images/ui-finetune/argilla_config.png)
```
Comment (sdiazlor) [inline ui-finetune/README.md:32]: This image looks to me a bit misleading due to the label in comparison to what you want to achieve/configure.

File: ui-finetune/README.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+pip install torch==2.0.1+cu118 torchvision==0.15.2+cu118 torchaudio==2.0.2 --extra-index-url https://download.pytorch.org/whl/cu118
+
+# Install other required libraries
+pip install -U autotrain-advanced
+pip install -U datasets transformers
+```
+
+### Select the algorithm
+
+There are countless fine-tuning algorithms for LLMs to chose from and many of the are supported by AutoTrain. We will work with the ORPO algorithm because it's simple to use and deliver significant improvements on base models.
```
Comment (sdiazlor) [inline ui-finetune/README.md]:
```suggestion
There are countless fine-tuning algorithms for LLMs to choose from, and many of them are supported by AutoTrain. We will work with the ORPO algorithm because it's simple to use and delivers significant improvements on base models. 
```

File: ui-finetune/README.md (hunk 2 above)
Comment (sdiazlor) [reply ui-finetune/README.md:80]: You can select it

Screenshot 2024-11-18 at 15 28 49

Comment (davidberenstein1957) [reply ui-finetune/README.md:80]: perhaps that photo is better to show?

Comment (burtenshaw) [reply ui-finetune/README.md:80]: No. There's an ORPO option. See the image below. I added a sentence to make this clearer.
